internal/querycoordv2/balance/balancer.go

```diff
@@ -0,0 +1,242 @@
+// Licensed to the LF AI & Data foundation under one
+// or more contributor license agreements. See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership. The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License. You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package balance
+
+import (
+	"go.uber.org/zap"
+
+	"github.com/milvus-io/milvus/internal/querycoordv2/meta"
+	"github.com/milvus-io/milvus/internal/querycoordv2/session"
+	"github.com/milvus-io/milvus/pkg/log"
+	"github.com/milvus-io/milvus/pkg/util/paramtable"
+	"github.com/samber/lo"
+)
+
+// Balance defines two things about balance impl details.
+// 1. what to balance.
+// for now, if node becomes stopping, or node has been transfer between resource groups.
+// segments/channels on it should be moved to other nodes, which and be achieved by balance;
+// also if segments/channels distribution on query node, we should trigger balance to do reassign.
+//
+// 2. how to balance.
+// balance should be triggered in order. the stopping node's balance should be triggered first,
+// which will recover the collection's load percents; then the transfer node operation's balance should be triggered,
+// to complete the transfer operation in async; the distribution balance should be triggered in last if auto balance is enabled.
+type Balance interface {
+	GetPolicy() BalancePolicy
+	BalanceReplica(replica *meta.Replica) ([]SegmentAssignPlan, []ChannelAssignPlan)
+}
+type Balancer struct {
+	nodeManager *session.NodeManager
+	dist        *meta.DistributionManager
+	meta        *meta.Meta
+	targetMgr   *meta.TargetManager
+
+	policy BalancePolicy
+}
+
+type nodeItem struct {
+	baseItem
+	nodeID int64
+}
+
+func newNodeItem(priority int, nodeID int64) nodeItem {
+	return nodeItem{
+		baseItem: baseItem{
+			priority: priority,
+		},
+		nodeID: nodeID,
+	}
+}
+
+type SegmentSkipBalanceFilter func(s *meta.Segment) bool
+
+func (b *Balancer) segmentRedundant(s *meta.Segment) bool {
+	return len(b.dist.SegmentDistManager.Get(s.GetID())) > 1
+}
+
+func (b *Balancer) segmentNotInTarget(s *meta.Segment) bool {
+	return !(b.targetMgr.GetHistoricalSegment(s.GetCollectionID(), s.GetID(), meta.CurrentTarget) != nil &&
+		b.targetMgr.GetHistoricalSegment(s.GetCollectionID(), s.GetID(), meta.NextTarget) != nil)
+}
+
```

Comment:
Can we just use:
(b.targetMgr.GetHistoricalSegment(s.GetCollectionID(), s.GetID(), meta.CurrentTarget) == nil  ||
 		b.targetMgr.GetHistoricalSegment(s.GetCollectionID(), s.GetID(), meta.NextTarget) == nil)
here?

Reply:
So the principle of moving one segment is that 'it must be both in current and next target', 
what if the stopping balance happened in this case? is this possible that one segment on stopping node will be discarded by this standard for the segment may not be in the next target?

Reply:
good catch!